### PR TITLE
update openapi and skd due to updated knative

### LINKED ISSUE #249: Upgrade openapi and python sdk once knative upgraded to v1beta1
/kind feature

**Describe the solution you'd like**
We need to upgrade openapi and python sdk once knative upgraded to v1beta1 (after the PR merged #238 ), since there are some changes in the `kfservice_types.go `

### PR DESCRIPTION
**Pull request by jinchihe**
**What this PR does / why we need it**:

Due to the knative is updated in the PR #251 , there are some changes in the knative and also has updates in kfservce_type.go, so we need to update the openapi, swagger file and python client.

**Which issue(s) this PR fixes** :
Fixes #249 

**Special notes for your reviewer**:

Here It is worth mentioning that: since the `URL` in the `KFServiceStatus` refer to`apis.URL` in `knative/pkg`, and that's aliased from `net.url`, we must to include the `net.url` while generete openapi, so I hacked to add `// +k8s:openapi-gen=true` in the `${GOROOT}/src/net/url/url.go` (No better way so far).

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/272)

### CODE REVIEW

**Comment by jinchihe** [comment]
/assign @yuzisun 
Please help to review once you get chance, thanks a lot!

File: sdk/docs/V1alpha1KFServiceStatus.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ Name | Type | Description | Notes
 **conditions** | [**list[KnativeCondition]**](KnativeCondition.md) | Conditions the latest available observations of a resource&#39;s current state. | [optional] 
 **default** | [**V1alpha1StatusConfigurationSpec**](V1alpha1StatusConfigurationSpec.md) |  | [optional] 
 **observed_generation** | **int** | ObservedGeneration is the &#39;Generation&#39; of the Service that was last processed by the controller. | [optional]
```
Comment (ellis-bigelow) [inline sdk/docs/V1alpha1KFServiceStatus.md]: I'm confused. I don't think we surface this in our API.

Comment (jinchihe) [reply sdk/docs/V1alpha1KFServiceStatus.md]: @ellis-bigelow The `KFServiceStatus` includes `duckv1beta1.Status` in the [kfservice_type.go](https://github.com/kubeflow/kfserving/blob/0ee00b3beec2ca9fb4eb6b64da1264ccb9b80914/pkg/apis/serving/v1alpha1/kfservice_types.go#L102), the `duckv1beta1.Status `includes [ObservedGeneration](https://github.com/knative/pkg/blob/2b848f71969c997809a2dbea8351eaadb23e9bc9/apis/duck/v1beta1/status_types.go#L59). So if we use the openapi-gen to generate, the `observed_generation` shown here, and in kfservice CRD define, also included [observedGeneration](https://github.com/kubeflow/kfserving/blob/0ee00b3beec2ca9fb4eb6b64da1264ccb9b80914/config/default/crds/serving_v1alpha1_kfservice.yaml#L296), as below:
```
[root@jinchi1 kfserving]# kubectl describe crd kfservices.serving.kubeflow.org |grep -i observedGeneration
              Description:  ObservedGeneration is the 'Generation' of the Service that was last processed by the controller.
```

Seems there is a problem, in the `kfservice_type.go`, we involved the `duckv1beta1.Status`, but only handled the `Conditions` and ignore `ObservedGeneration`, seems also we need to handle the `ObservedGeneration`? or update the `kfservice_type.go` to just involve `apis.Conditions`？WDYT?

Comment (yuzisun) [reply sdk/docs/V1alpha1KFServiceStatus.md]: ye I think it is good to handle this, knative service has that.

File: hack/update-openapigen.sh [hunk 2]
```diff
@@ -22,17 +22,38 @@ if [ -z "${GOPATH:-}" ]; then
...
-sed -i '/^type VolatileTime struct {/i // +k8s:openapi-gen=true' vendor/github.com/knative/pkg/apis/volatile_time.go
+if [ -z "${GOROOT:-}" ]; then
+    export GOROOT=$(go env GOROOT)
+fi
+
+# Below hacks are to add "// +k8s:openapi-gen=true" for some dependencies.
+# The knative related hacking can be removed once upgraded (knative/pkg/pull/510).
+knative_condition_file="vendor/github.com/knative/pkg/apis/condition_types.go"
+knative_volatile_file="vendor/github.com/knative/pkg/apis/volatile_time.go"
+knative_url_file="vendor/github.com/knative/pkg/apis/url.go"
```
Comment (yuzisun) [inline hack/update-openapigen.sh]: should we create our own url type ? Looks it generates a bunch of fields that we do not need.

**Comment by ellis-bigelow** [comment]
Perhaps relying on the conditions instead is better. I want to be able to
use the conditionset that they provide. I don't see how observed generation
is useful. Is it worth reaching out to knative and seeing what it's for?

**Comment by jinchihe** [comment]
@yuzisun @ellis-bigelow Checked in knative, the `observedGeneration` is is the 'Generation' of the Service that was last processed by the controller. For example, I created the `kservice`, the `Genareate` and `observedGeneration` are `1`, and applied to changed something, the `Genareate` and `observedGeneration` are `2`.

```
[root@jinchi1 ~]# kubectl get ksvc helloworld-go -o yaml
apiVersion: serving.knative.dev/v1alpha1
kind: Service
metadata:
  ...
  creationTimestamp: 2019-08-05T01:42:02Z
  generation: 2
  name: helloworld-go
   ...

status:
   ...
  observedGeneration: 2
   ...
```

Personally I suggest to keep the current implement, since no impacts and not sure if that's useful for `kfserving` in the future although it doesn't seem very useful right now. Comments?

File: hack/update-openapigen.sh (hunk 2 above)
Comment (jinchihe) [reply hack/update-openapigen.sh]: Hi @yuzisun That's needed, Yes, we can use that `URL` struct directly, but in the URL struct, there is a [Userinfo](https://github.com/golang/go/blob/a2f5d644d3a2903a82e527e3bca2c4bdff1d0750/src/net/url/url.go#L359) that's linked to another struch in the file. If not creating the url type, there will be have import traceback [here](https://github.com/kubeflow/kfserving/blob/8ddc3f027187134f241888c0dda1572ff06533ac/sdk/kfserving/models/knative_url.py#L19) and no link for the document [KnativeURL.md](https://github.com/kubeflow/kfserving/blob/8ddc3f027187134f241888c0dda1572ff06533ac/sdk/docs/KnativeURL.md)

**Comment by yuzisun** [comment]
> Perhaps relying on the conditions instead is better. I want to be able to use the conditionset that they provide. I don't see how observed generation is useful. Is it worth reaching out to knative and seeing what it's for?
> […](#)

we actually need to use knative status duck type otherwise the condition propagation won't work properly.  I am not sure what's involved to just rely on conditions.

**Comment by jinchihe** [comment]
> we actually need to use knative status duck type otherwise the condition propagation won't work properly. I am not sure what's involved to just rely on conditions.

@yuzisun @ellis-bigelow if so, I created the new ticket #278 to trace handling the `observedGeneration`. Let's keep that as implemented of SDK and CRD defination, and address that in seprated ticket. Thanks.

**Comment by ellis-bigelow** [comment]
Cool. Sounds like a good conclusion for now

**Comment by yuzisun** [comment]
@ellis-bigelow @johnugeorge @jinchihe by introducing `apis.URL` to status, we have to create this hack since that struct is deep in go standard library(https://github.com/golang/go/blob/a2f5d644d3a2903a82e527e3bca2c4bdff1d0750/src/net/url/url.go#L359) and we can not add `openapi` annotation, I'd suggest create our simplified URL struct without unnecessary fields such as `userinfo` which pull in a lot of other unrelated fields. WDYT ?

**Comment by jinchihe** [comment]
> @ellis-bigelow @johnugeorge @jinchihe by introducing `apis.URL` to status, we have to create this hack since that struct is deep in go standard library(https://github.com/golang/go/blob/a2f5d644d3a2903a82e527e3bca2c4bdff1d0750/src/net/url/url.go#L359) and we can not add `openapi` annotation, I'd suggest create our simplified URL struct without unnecessary fields such as `userinfo` which pull in a lot of other unrelated fields. WDYT ?

@yuzisun After investigation, I think that's better to rollback to `string` for URL since if we ceate our URL struct, that's hard to assign with net.url (cannot sure no issue in the future). And there is a [String function](https://github.com/knative/pkg/blob/d3c17af1d97a4dd7f0167628df312f1e9889f444/apis/url.go#L69) in the knative.dev/pkg/apis.URL, we can use that and transfer to string and kfservice.

Since the topic is related with other issue than updating sdk, will send separated PR to update that. Thanks.

**Comment by jinchihe** [comment]
/hold
For two things: 
1. waiting for #289 merged, and update the PR accordingly.
2. the knative has been upgraded to 0.8, will handle SDK accordingly.

**Comment by yuzisun** [comment]
@jinchihe you can remove all the hacks now :)

**Comment by jinchihe** [comment]
> @jinchihe you can remove all the hacks now :)

@yuzisun Yes, you are right!

**Comment by jinchihe** [comment]
@yuzisun Sorry.. I made a mistake while trying to rebase, and caused the PR closed, frustrated basing due to changed laptop.. I will try the send another PR to upgrade this.

**Pull request closed without merging**